File: Problem-Set-Volumes-(100...1999)/Volume-1-(100-199)/103 - Stacking Boxes/answer.py
def can_nest(box_a, box_b):
    return all(dim_a < dim_b for dim_a, dim_b in zip(box_a['dimensions'], box_b['dimensions']))

def find_longest_nesting_sequence(lines):
    results = []
    i = 0
    while i < len(lines):
        k, n = map(int, lines[i].split())
        i += 1
        boxes = []
        
        for j in range(k):
            dimensions = list(map(int, lines[i].split()))
            boxes.append({'index': j + 1, 'dimensions': sorted(dimensions)})
            i += 1
            
        boxes.sort(key=lambda box: box['dimensions'])
        
        graph = [[] for _ in range(k)]
        for u in range(k):
            for v in range(u + 1, k):
                if can_nest(boxes[u], boxes[v]):
                    graph[u].append(v)
        
        dp = [1] * k
        prev = [-1] * k
        
        for u in range(k):
            for v in graph[u]:
                if dp[u] + 1 > dp[v]:
                    dp[v] = dp[u] + 1
                    prev[v] = u
                elif dp[u] + 1 == dp[v] and u < prev[v]:
                    prev[v] = u
        
        max_length = 0
        max_index = -1
        for u in range(1, k):
            if dp[u] >= max_length:
                max_length = dp[u]
                max_index = u
        
        sequence = []
        while max_index != -1:
            sequence.append(boxes[max_index]['index'])
            max_index = prev[max_index]
        
        results.append(str(max_length))
        results.append(" ".join(map(str, reversed(sequence))))
    
    return "\n".join(results)

def main():
    import sys
    lines = []
    print("Enter input (end with an empty line):")
    for line in sys.stdin:
        if line.strip() == "":
            break
        lines.append(line.strip())
    
    output = find_longest_nesting_sequence(lines)
    print(output)

if __name__ == "__main__":
    main()

File: Problem-Set-Volumes-(100...1999)/Volume-1-(100-199)/103 - Stacking Boxes/answer.js
const readline = require("readline");

function canNest(boxA, boxB) {
  return boxA.dimensions.every((dim, index) => dim < boxB.dimensions[index]);
}
function findLongestNestingSequence(lines) {
  let i = 0;
  const results = [];

  while (i < lines.length) {
    const [k, n] = lines[i++].split(" ").map(Number);
    const boxes = [];
    for (let j = 0; j < k; j++) {
      const dimensions = lines[i++].split(" ").map(Number);
      boxes.push({
        index: j + 1,
        dimensions: dimensions.slice().sort((a, b) => a - b),
      });
    }
    boxes.sort((a, b) => {
      for (let i = 0; i < a.dimensions.length; i++) {
        if (a.dimensions[i] !== b.dimensions[i]) {
          return a.dimensions[i] - b.dimensions[i];
        }
      }
      return 0;
    });
    const graph = Array(k)
      .fill(0)
      .map(() => []);
    for (let u = 0; u < k; u++) {
      for (let v = u + 1; v < k; v++) {
        if (canNest(boxes[u], boxes[v])) {
          graph[u].push(v);
        }
      }
    }
    const dp = Array(k).fill(1);
    const prev = Array(k).fill(-1);
    for (let u = 0; u < k; u++) {
      for (const v of graph[u]) {
        if (dp[u] + 1 > dp[v]) {
          dp[v] = dp[u] + 1;
          prev[v] = u;
        } else if (dp[u] + 1 === dp[v] && u < prev[v]) {
          prev[v] = u;
        }
      }
    }
    let maxLength = 0;
    let maxIndex = -1;
    for (let u = 1; u < k; u++) {
      if (dp[u] >= maxLength) {
        maxLength = dp[u];
        maxIndex = u;
      }
    }
    const sequence = [];
    while (maxIndex !== -1) {
      sequence.push(boxes[maxIndex].index);
      maxIndex = prev[maxIndex];
    }

    results.push(`${maxLength}`);
    results.push(sequence.reverse().join(" "));
  }

  return results.join("\n");
}

const main = async () => {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
  });
  const lines = [];
  console.log("Enter input (end with an empty line):");
  for await (const line of rl) {
    if (line.trim() === "") break;
    lines.push(line.trim());
  }
  const output = findLongestNestingSequence(lines);
  console.log(output);
  rl.close();
};
main();
